### PR TITLE
[RzIL] Add Validator (Type-checker)

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Well-typed programs cannot go wrong.
  -- Robin Milner

This adds the IL validator, which performs static type-checking of both pure and effect ops among other checks. In particular, assuming the validator is correctly implemented, if it considers an op to be valid under some context, evaluating the op in the vm will never yield a runtime error, that is, an error where the vm itself errors, not an expected error state of the code being executed.
In our case, this includes for example:
* Any kind of type error: Conditions not being bool, bitvector sizes not matching, ...
* Variables not being available when they are accessed
* Using local variables with multiple different types in a single effect
* etc.

Any code that we lift must obey these rules. Thus, any analysis can rely on it.
The plan for this is to use the validator primarily in testing, development of lifters and for IL code coming from the outside.
If our lifting code is covered well enough by tests using the validator, we can omit the validation at runtime.

The only ops that do not have well-defined validation yet are `blk` and `goto` since their semantics, in particular regarding label handling are still a bit vague.
This also removes the `concat` and `unk` ops since they are unimplemented and not needed.

**Test plan**

See the unit tests.

### CODE REVIEW

File: librz/il/il_export.c
Comment (XVilka) [inline librz/il/il_export.c:832]: Add a brief Doxygen please

File: librz/il/il_export.c
Comment (XVilka) [inline librz/il/il_export.c]: Same

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:32]: `pc_len` - could it be zero? Are there any invalid values? Maybe add checks for them?

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c]: Please add `RZ_NONNULL` to the `name`

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:58]: Check for `hts == NULL`.

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c]: `user` is being dereferenced later, please add `RZ_NONNULL`

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c]: Same.

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c]: Same

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:207]: Same

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:244]: You could use grouping feature by Doxygen with `\@{`, etc

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:264]: Broken indentation?

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:274]: Maybe a newline here? To make it more readable

File: librz/util/str_trim.c
Comment (XVilka) [inline librz/util/str_trim.c]: Add then `RZ_NONNULL` here too.

Comment (XVilka) [reply librz/util/str_trim.c]: And ownership mark of the return, `RZ_BORROW`

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:610]: Same, could use Doxygen grouping.

File: librz/il/il_validate.c
Comment (XVilka) [inline librz/il/il_validate.c:406]: For me, it's obvious that we are dealing with binary ops, then unary ops, etc. But for new devs it's probably worth to add a brief description for those validation functions.

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall looks very good. LGTM once you address my nitpicks.

File: librz/il/il_validate.c
Comment (thestr4ng3r) [reply librz/il/il_validate.c:32]: Should indeed not be zero.

File: librz/il/il_validate.c
Comment (thestr4ng3r) [reply librz/il/il_validate.c:264]: clang-format's fault

**Comment by thestr4ng3r** [comment]
I don't see what is supposed to be red there.

**Review by XVilka — APPROVED** [review]

**Comment by wargio** [comment]
i wanted to check this before merging

**Pull request merged**
